### ISSUE TITLE
How to train MLP model for RGBXY to Gx, GY mapping

### ISSUE THREAD

**Issue by vocdex** [leo9344/Implementation-of-GelSight#1]
Thanks for such detailed instructions.
I am trying to replicate the steps in the paper. The authors have trained a small MLP network (5-32-32-32-2). I guess you also have collected data for this network and probably trained it.
Would it be possible to get the script for model training? Did you concatenate RGBXYGxGy values of each image into one huge CSV file? 
If I am not mistaken, GelSight images are 320x240, so we should get 320x240 row and 5 column array for each image, right? How many images were enough for the training in your case?
Thanks in advance!

**Comment by leo9344** [comment]
Thank you for trying our project!

For training the model: [https://github.com/leo9344/Implementation-of-GelSight/blob/main/color2grad.py](https://github.com/leo9344/Implementation-of-GelSight/blob/main/color2grad.py)

We concatenated RGBXY values for input and GxGy for target of each images. It's in **line 65** of [https://github.com/leo9344/Implementation-of-GelSight/blob/main/color2grad.py](https://github.com/leo9344/Implementation-of-GelSight/blob/main/color2grad.py).

And the data we used for training and testing is in **./data/train.csv** and **./data/test.csv**. Original (unsplit) CSVs are in **./data/XYBGR.csv** and **./data/XYGxGy.csv**. These data are collected by @Mrliwanlin patiently and carefully. 

Our images are 100x100 and that's different from original Gelsight images. So you can get 100x100x5 array for each image in those CSVs. But you can collect 320x240 images and train a MLP as well. :) 

We collected 30 images approximately and we had **a severe overfitting problem**. That's because most of our training images (https://github.com/leo9344/Implementation-of-GelSight/blob/main/data/BallCali_wanlin.zip/data/*.png) are captured when the sensor is attatched to a single, measured ball. So if you have tested our model, you may found whatever the input is, the model is giving a ball-like output. So the **mapping table may be more useful for now**. 

I'm sorry to say that I couldn't follow this project because some personal reasons. And @Mrliwanlin is still trying to improve the performance of this sensor. You can also email him to get fresher news. 😄 

Again, thanks for testing our project and hope you can get a better model. If you have any further questions, please feel free to ask.

**Comment by vocdex** [comment]
Thank you for such an amazing response. Now, everything is clear.
It seems like the MLP model is not sufficient for non-sphere objects. I wonder how Gelsight Wedge authors were able to stick with this approach in their paper, given its pitfalls. Maybe, there's a way to get the ground truth values for non-sphere objects and add these objects to training. I will look into it.

**Comment by Mrliwanlin** [comment]
Hi vocdex,

Actually, the MLP works for the non-sphere objects and you could just follow the steps as @leo9344 posted, and see the results.

**Comment by vocdex** [comment]
Hi @Mrliwanlin,
Thanks for your suggestion. I was wondering if you had compared the performance of the lookup table to that of MLP. Also, could you please list the steps to produce a lookup table for GelSight sensor? 
Thanks

**Comment by Mrliwanlin** [comment]
From my perspective, both methods work well and the advantage of MLP is that it provides a smoother reconstruction results but it lacks of details of the object. The lookup table would provide more detailed reconstruction but have more noises. 

The steps of lookup table when using a sphere to calibrate could be concluded as follows:
1. Outline the contour of the sphere in the image;
2. Build an ideal hemisphere based on the contour and then calculate the corresponding gradients Gx, Gy in X and Y direction;
3. Mapping the R, G, B channel values of the pixels of the hemisphere to the Gx and Gy to establish the table.

**Comment by vocdex** [comment]
Thank you. I will do both and compare once again.
